frame/staking/src/offchain_election.rs

```diff
@@ -0,0 +1,116 @@
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helpers for offchain worker election.
+
+use crate::{Call, CompactAssignments, Module, SessionInterface, Trait};
+use codec::Encode;
+use frame_system::offchain::{SubmitUnsignedTransaction};
+use sp_phragmen::{reduce, ExtendedBalance, PhragmenResult, StakedAssignment};
+use sp_std::{prelude::*, cmp::Ordering};
+use sp_runtime::{RuntimeAppPublic, RuntimeDebug};
+
+#[derive(RuntimeDebug)]
+pub(crate) enum OffchainElectionError {
+	/// No signing key has been found on the current node that maps to a validators. This node
+	/// should not run the offchain election code.
+	NoSigningKey,
+	/// Phragmen election returned None. This means less candidate that minimum number of needed
+	/// validators were present. The chain is in trouble and not much that we can do about it.
+	FailedElection,
+}
+
+/// Compares two sets of phragmen scores based on desirability and returns true if `that` is
+/// better `this`.
+///
+/// Evaluation is done in a lexicographic manner.
+pub(crate) fn is_score_better(this: [ExtendedBalance; 3], that: [ExtendedBalance; 3]) -> bool {
+	match that
+		.iter()
+		.enumerate()
+		.map(|(i, e)| e.cmp(&this[i]))
+		.collect::<Vec<Ordering>>()
+		.as_slice()
+	{
+		[Ordering::Greater, _, _] => true,
+		[Ordering::Equal, Ordering::Greater, _] => true,
+		[Ordering::Equal, Ordering::Equal, Ordering::Less] => true,
+		_ => false,
+	}
+}
+
+/// The internal logic of the offchain worker of this module.
+pub(crate) fn compute_offchain_election<T: Trait>() -> Result<(), OffchainElectionError> {
+	let validator_keys = T::SessionInterface::keys::<T::KeyType>();
+	let local_keys = T::KeyType::all();
+
+	if let Some((index, ref pubkey)) = local_keys
+		.into_iter()
+		.find_map(|k|
+			validator_keys
+				.iter()
+				.enumerate()
+				.find_map(|(index, (_acc, maybe_vk))|
+					maybe_vk.as_ref()
+						.and_then(|vk| if *vk == k { Some((index, vk)) } else { None })
+				)
+		) {
+			// k is a local key who is also among the validators.
+			let PhragmenResult {
+				winners,
+				assignments,
+			} = <Module<T>>::do_phragmen().ok_or(OffchainElectionError::FailedElection)?;
+
+			// convert winners into just account ids.
+			let winners: Vec<T::AccountId> = winners.into_iter().map(|(w, _)| w).collect();
+
+			// convert into staked. This is needed to be able to reduce.
+			let mut staked: Vec<StakedAssignment<T::AccountId>> = assignments
+				.into_iter()
+				.map(|a| a.into_staked::<_, _, T::CurrencyToVote>(<Module<T>>::slashable_balance_of))
+				.collect();
+
+			// reduce the assignments. This will remove some additional edges.
+			reduce(&mut staked);
```

Comment:
Right now, no although, your offchain code will run it. In general though, it is incentivised because you will pay less transaction fees (not the case for validators.)

Reply:
So for validator node I should skip it, but if the cost of the function is not to big, I would rather run a standard runtime and do not care so it should be fine.